### DISCUSSION TITLE
How can i couple the displacement between two mesh?

### DISCUSSION THREAD

**Discussion by HoseongJeong** [kinnala/scikit-fem#1171] [General]
Dear scikit-fem. 
Thanks for all the comments that I made last time!

I am trying to model an embedded rabar in concrete , where the rebar is 1D element and the concrete is 3D element. (They are separately meshed.)

The displacement of rebar should be equal to the concrete surrounding the rebar. 

In this case, how can I couple the displacement between the surrounding concrete and rebar? 

Once again,
Very thanks all the staff and users of scikit-fem!

**Comment by kinnala** [comment]
I think we don’t have a good tool for this purpose currently.

There is Basis for integrating over 3D elements, there is FacetBasis for integrating over 2D elements, but there is nothing like EdgeBasis for integrating over 1D elements.

I think there are no real blockers to prevent implementing such thing but it would require modifications at least in
- Mesh3D to support tagging of edge subsets
- MappingAffine to support local-to-global mapping for edges
- implementation of EdgeBasis and its testing

This is not a small feat but definitely doable I think.

**Reply by HoseongJeong** [reply]
Would you please put this functions in the next update? It would be very helpful for the users like researcher of composite material to perform structural analysis! I will make sure to promote scikit-fem by citing my research paper!!!

**Reply by kinnala** [reply]
I will now quickly try to make a prototype of the above to see how much work is required.

**Reply by kinnala** [reply]
See #1172 for a prototype. Assuming it works, it wasn't that bad, but more testing is definitely required. I currently only check that integration over edges seems to work correctly.

Note that there is only basic support for first-order tetrahedral mesh and hexahedral mesh is not supported because it would require doing similar changes in `MappingIsoparametric`. Those changes might be more challenging than for `MappingAffine` because the local-to-global mapping is not linear for hexahedral meshes.

I cannot promise that this will be useful for your work in the near future since I am busy with teaching and research. If you want, you can also continue from there if you find the work useful.

**Reply by HoseongJeong** [reply]
Really appreciate it!!! I will try this code! 
You are always kind and supportive person 💗 💛.
